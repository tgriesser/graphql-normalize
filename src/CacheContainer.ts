import produce, { Immutable } from 'immer';
import type { CacheShape } from './cache';

export class CacheContainer {
  #state: Immutable<CacheShape>;
  #subscribers: Function[] = [];

  constructor(initialState?: CacheShape) {
    this.#state = produce({ fields: {}, operations: {} }, (root) => initialState ?? root);
  }

  getState() {
    return this.#state;
  }

  subscribe(fn: (state: Immutable<CacheShape>) => void) {
    this.#subscribers.push(fn);
    return () => {
      this.#subscribers.splice(this.#subscribers.indexOf(fn), 1);
    };
  }
}
