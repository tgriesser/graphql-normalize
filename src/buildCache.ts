import type { FormattedExecutionResult } from 'graphql';
import { walk } from './walk';
import type { NormalizedDoc } from './extensionShape';

/**
 * Write values into the cache, given the values, the existing cache, and the
 */
export function buildCache(result: FormattedExecutionResult, normalizingMeta: NormalizedDoc) {
  let pathPrefix = '';
  let currentKey;
  let currentKeyStack = [];
  let currentWriteKey = undefined;
  let currentWriteKeyStack = [];

  const selectionKeys = Object.keys(normalizingMeta.selectionSet);

  function walkUntilList(val: any) {
    walk(val, {
      enter(node) {
        const key = node.path.join('.');
        const metaDef = normalizingMeta.selectionSet[key];
        console.log(key, metaDef);
        if (metaDef.list) {
          return false;
        }
      },
      leave() {
        //
      },
    });
  }

  walkUntilList(result);
}
