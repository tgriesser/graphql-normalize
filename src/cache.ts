import type { NormalizeMetaShape } from './metadataShapes';

type ArgsKey = string;

export interface CacheShape {
  operations: {
    // The metadata for each query
    [hash: string]: {
      meta: NormalizeMetaShape;
    };
  };
  fields: {
    [key: string]: any;
  };
}
