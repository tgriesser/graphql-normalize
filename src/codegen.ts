import { NormalizeCodegenPlugin } from './codegen/plugin.js';
export type { TypePolicies } from './codegen/getCacheKey.js';
export type { NormalizeMetaShape, NormalizedDocShape } from './metadataShapes.js';

export { NormalizeCodegenPlugin } from './codegen/plugin.js';
export { generateNormalizedMetadataForDocs, generateNormalizedMetadata } from './codegen/generateNormalizedMetadata.js';
export { generateNormalizedOperations, generateNormalizedOperation } from './codegen/generateNormalizedOperation.js';

export { NormalizeCodegenPlugin as plugin };

export default {
  plugin: NormalizeCodegenPlugin,
};
