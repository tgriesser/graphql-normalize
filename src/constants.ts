export const __typename = '__typename' as const;
