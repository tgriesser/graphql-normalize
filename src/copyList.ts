interface CopyListConfig {
  /**
   * The source list of values we're looking to copy
   */
  source: Array<any>;
  /**
   * The target list the source will be copied into,
   * used for equality checking
   */
  target: Array<any>;
  /**
   * Whether we mutate the target, or create a new list
   * when modifications occur
   * @default true
   */
  mutate?: boolean;
  /**
   * Depth of items to copy. Once we've hit the depth,
   * even if there are further arrays, we'll use equality checking
   */
  depth: number;
  currentDepth?: number;
  indexes?: number[];
  isEqual: (a: any, b: any) => boolean;
  /**
   * Called with the existing & new values, as well as the indexes
   * of the current array
   */
  onValue: (existingValue: any, newValue: any, indexes: number[]) => any;
}

/**
 * Recursively & mutatively copies a list, up to a specified depth.
 *
 * The list mutation is intentionally mutable, as this is intended to be
 * executed within an "immer" produce function, which deals with immutability
 */
export function copyList(config: CopyListConfig) {
  const { source, target, depth, currentDepth = 1, onValue, indexes = [], isEqual } = config;
  let modified = false;
  if (source.length < target.length) {
    modified = true;
    target.length = source.length;
  }
  for (let i = 0; i < source.length; i++) {
    const value = source[i];
    if (value === null) {
      if (target[i] !== null) {
        target[i] = null;
        modified = true;
      }
      continue;
    }
    if (currentDepth < depth) {
      const { list, modified: innerModified } = copyList({
        ...config,
        source: value,
        target: target[i] ?? [],
        currentDepth: currentDepth + 1,
        indexes: indexes.concat(i),
      });
      if (innerModified) modified = true;
      target[i] = list;
    } else {
      const newValue = onValue(target[i], value, indexes.concat(i));
      if (!isEqual(newValue, target[i])) {
        target[i] = newValue;
        modified = true;
      }
    }
  }
  return { list: target, modified };
}
