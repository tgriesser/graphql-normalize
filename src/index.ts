import { graphqlNormalize } from './graphqlNormalize.js';
export { stringifyVariables } from './stringifyVariables.js';
export { MissingFieldError } from './errors.js';
export type { graphqlNormalizeOptions, NormalizeMetaShape } from './graphqlNormalize.js';
export { graphqlNormalize };
export default graphqlNormalize;
