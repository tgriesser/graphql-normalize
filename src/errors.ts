import { NO_ARGS } from './constants.js'

export class MissingFieldError {
  constructor(path: Array<string | number>) {
    throw new Error(`Missing field for: ${path.filter((p) => p !== NO_ARGS)}`)
  }
}
