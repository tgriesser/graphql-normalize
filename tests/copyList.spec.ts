import { describe, expect, it } from 'vitest';
import { copyList } from '../src/copyList';

describe('copyList', () => {
  it('copies list from source to destination', () => {
    const source = [
      [['a', 'b'], ['c']],
      ['d', ['e', 'f']],
    ];
    const destination: any[] = [];
    const result = copyList({
      source,
      target: destination,
      depth: 3,
      onValue: (a, b) => b,
      isEqual: (a, b) => a === b,
    });
    expect(result.modified).toEqual(true);
    expect(result.list).toMatchSnapshot();
  });

  it('doesnt mutate the list if nothing has changed', () => {
    const source = [
      [['a', 'b'], ['c']],
      ['d', ['e', 'f']],
    ];
    const destination: any[] = [
      [['a', 'b'], ['c']],
      ['d', ['e', 'f']],
    ];
    const result = copyList({
      source,
      target: destination,
      depth: 3,
      onValue: (a, b) => b,
      isEqual: (a, b) => a === b,
    });
    expect(result.modified).toEqual(false);
    expect(result.list).toEqual(destination);
  });
});
