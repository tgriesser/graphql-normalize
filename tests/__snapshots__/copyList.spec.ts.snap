// Vitest Snapshot v1

exports[`copyList > copies list from source to destination 1`] = `
[
  [
    [
      "a",
      "b",
    ],
    [
      "c",
    ],
  ],
  [
    [
      "d",
    ],
    [
      "e",
      "f",
    ],
  ],
]
`;
